adding split_left/split_right for reducing memory usage

tiye opened the pull request:
related context https://github.com/aslpavel/fingertree-rs/issues/2#issuecomment-978996502 .

tiye committed:
adding split_left/split_right for reducing memory usage

tiye commented on .rustfmt.toml:
```diff
@@ -0,0 +1 @@
+tab_spaces = 4
```
ensures VS Code..

tiye commented on src/lib.rs:
```diff
@@ -219,6 +219,39 @@ where
+        } else {
+            Self::new()
+        }
+    }
```
passing `split_right` around looks dazzling, this is the parts I'm not sure.

tiye commented:
`split_right` is slower compare to `split_left` due to calling of `.measure()`s. For a while I thought maybe we need a `split_from_right` but there is still costs in reversed iterating/slicing of Digits.

```
Benchmarking split/position/211: Collecting 100 samples in estimated 5.0005 s (2.2M iter
  split/position/211      time:   [2.0548 us 2.0740 us 2.0945 us]
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) high mild
  10 (10.00%) high severe
Benchmarking split/position/384: Collecting 100 samples in estimated 5.0034 s (2.3M iter
  split/position/384      time:   [2.1363 us 2.1717 us 2.2280 us]
Found 10 outliers among 100 measurements (10.00%)
  7 (7.00%) high mild
  3 (3.00%) high severe
Benchmarking split/position/557: Collecting 100 samples in estimated 5.0100 s (2.0M iter
  split/position/557      time:   [2.2377 us 2.2670 us 2.2981 us]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
Benchmarking split/position/730: Collecting 100 samples in estimated 5.0099 s (2.4M iter
  split/position/730      time:   [2.0459 us 2.0596 us 2.0780 us]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe
Benchmarking split/position/903: Collecting 100 samples in estimated 5.0032 s (2.5M iter
  split/position/903      time:   [1.8699 us 1.9071 us 1.9642 us]
Found 12 outliers among 100 measurements (12.00%)
  2 (2.00%) high mild
  10 (10.00%) high severe

Benchmarking split_left/position/211: Collecting 100 samples in estimated 5.0049 s (4.8M
  split_left/position/211 time:   [1.0481 us 1.0554 us 1.0634 us]
Found 4 outliers among 100 measurements (4.00%)
  4 (4.00%) high mild
Benchmarking split_left/position/384: Collecting 100 samples in estimated 5.0038 s (5.4M
  split_left/position/384 time:   [953.89 ns 1.0002 us 1.0600 us]
Found 14 outliers among 100 measurements (14.00%)
  8 (8.00%) high mild
  6 (6.00%) high severe
Benchmarking split_left/position/557: Collecting 100 samples in estimated 5.0038 s (5.1M
  split_left/position/557 time:   [1.1018 us 1.2474 us 1.4536 us]
Found 15 outliers among 100 measurements (15.00%)
  6 (6.00%) high mild
  9 (9.00%) high severe
Benchmarking split_left/position/730: Collecting 100 samples in estimated 5.0044 s (5.0M
  split_left/position/730 time:   [926.98 ns 947.53 ns 974.26 ns]
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe
Benchmarking split_left/position/903: Collecting 100 samples in estimated 5.0014 s (5.5M
  split_left/position/903 time:   [773.03 ns 798.05 ns 828.16 ns]
Found 12 outliers among 100 measurements (12.00%)
  6 (6.00%) high mild
  6 (6.00%) high severe

Benchmarking split_right/position/211: Collecting 100 samples in estimated 5.0009 s (5.2
  split_right/position/211
                        time:   [980.87 ns 1.0005 us 1.0253 us]
Found 13 outliers among 100 measurements (13.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  9 (9.00%) high severe
Benchmarking split_right/position/384: Collecting 100 samples in estimated 5.0023 s (3.9
  split_right/position/384
                        time:   [1.2055 us 1.2508 us 1.3040 us]
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) high mild
  7 (7.00%) high severe
Benchmarking split_right/position/557: Collecting 100 samples in estimated 5.0053 s (4.3
  split_right/position/557
                        time:   [1.1336 us 1.1366 us 1.1396 us]
Found 11 outliers among 100 measurements (11.00%)
  2 (2.00%) low mild
  3 (3.00%) high mild
  6 (6.00%) high severe
Benchmarking split_right/position/730: Collecting 100 samples in estimated 5.0044 s (4.3
  split_right/position/730
                        time:   [1.1690 us 1.1722 us 1.1754 us]
Found 14 outliers among 100 measurements (14.00%)
  1 (1.00%) low severe
  2 (2.00%) low mild
  6 (6.00%) high mild
  5 (5.00%) high severe
Benchmarking split_right/position/903: Collecting 100 samples in estimated 5.0052 s (4.7
  split_right/position/903
                        time:   [1.0690 us 1.0709 us 1.0734 us]
Found 14 outliers among 100 measurements (14.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  12 (12.00%) high severe
```

aslpavel commented:
Looks great, couple comments though:
- I think there is no much use in having `Digit::split_{right|left}`  since `Digit::split` does not allocate anything anyway, and computing measure for it is super-fast as it is limited by four elements. I've ran some test it does not effect performance.
- As for the `right_split` if would prefer if it would return not the next  measure but rather the measure of the tree node if it was not thrown away. That would make code look very similar to `Tree::split`
- Since split on digit returns slices you can define new impl in `measure.rs`
```rust
impl<'a, T: Measured> Measured for &[T] {
    type Measure = T::Measure;

    fn measure(&self) -> Self::Measure {
        self.iter()
            .fold(T::Measure::unit(), |acc, val| acc.join(&val.measure()))
    }
}
```

tiye committed:
unify digit splitting methods as review comment

tiye commented:
> I've ran some test it does not effect performance.

Indeed. I changed updated the code, only got `< 4%` more cost compared previous code.

```
split/position/211      time:   [2.0207 us 2.0352 us 2.0535 us]
                        change: [-24.984% -16.786% -8.3930%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) high mild
  11 (11.00%) high severe
split/position/384      time:   [2.1290 us 2.1328 us 2.1374 us]
                        change: [-2.9448% -1.0737% +0.2787%] (p = 0.23 > 0.05)
                        No change in performance detected.
Found 14 outliers among 100 measurements (14.00%)
  4 (4.00%) low mild
  3 (3.00%) high mild
  7 (7.00%) high severe
split/position/557      time:   [2.1344 us 2.1404 us 2.1484 us]
                        change: [-5.5715% -4.3774% -3.1699%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) high mild
  11 (11.00%) high severe
split/position/730      time:   [2.0328 us 2.0374 us 2.0418 us]
                        change: [-1.0124% -0.3132% +0.4635%] (p = 0.42 > 0.05)
                        No change in performance detected.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe
split/position/903      time:   [1.8531 us 1.8574 us 1.8616 us]
                        change: [-6.0016% -2.9199% -0.4638%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe

split_left/position/211 time:   [1.0457 us 1.0495 us 1.0529 us]
                        change: [-1.2582% -0.4570% +0.2849%] (p = 0.25 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) low mild
  3 (3.00%) high mild
  3 (3.00%) high severe
split_left/position/384 time:   [924.12 ns 928.68 ns 933.73 ns]
                        change: [-15.862% -10.187% -5.2053%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
  7 (7.00%) high severe
split_left/position/557 time:   [972.44 ns 976.77 ns 981.42 ns]
                        change: [-27.814% -20.388% -12.907%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe
split_left/position/730 time:   [914.61 ns 932.36 ns 952.71 ns]
                        change: [-8.1944% -6.0987% -4.0069%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 14 outliers among 100 measurements (14.00%)
  5 (5.00%) high mild
  9 (9.00%) high severe
split_left/position/903 time:   [765.44 ns 766.74 ns 768.09 ns]
                        change: [-5.7146% -3.4511% -1.3902%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) low severe
  1 (1.00%) low mild
  2 (2.00%) high mild
  3 (3.00%) high severe

split_right/position/211
                        time:   [958.68 ns 961.51 ns 964.30 ns]
                        change: [-4.6944% -3.0025% -1.5383%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) low severe
  2 (2.00%) low mild
  1 (1.00%) high mild
  3 (3.00%) high severe
split_right/position/384
                        time:   [1.2131 us 1.2453 us 1.2811 us]
                        change: [-6.5382% -2.1777% +2.2225%] (p = 0.34 > 0.05)
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  4 (4.00%) high mild
  6 (6.00%) high severe
split_right/position/557
                        time:   [1.1571 us 1.1625 us 1.1690 us]
                        change: [+2.3661% +3.9256% +6.0940%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 13 outliers among 100 measurements (13.00%)
  2 (2.00%) low mild
  5 (5.00%) high mild
  6 (6.00%) high severe
split_right/position/730
                        time:   [1.2060 us 1.2088 us 1.2114 us]
                        change: [+2.0233% +2.9740% +3.9107%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 6 outliers among 100 measurements (6.00%)
  3 (3.00%) high mild
  3 (3.00%) high severe
split_right/position/903
                        time:   [1.0897 us 1.0924 us 1.0954 us]
                        change: [+0.6780% +1.6687% +2.6705%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) high mild
  4 (4.00%) high severe
```

well, not stable, but still `<4%` costs:

```
warning: `fingertrees` (bench "criterion") generated 6 warnings
    Finished bench [optimized] target(s) in 7.76s
     Running unittests (target/release/deps/criterion-07ed78d4fef3ab04)
Gnuplot not found, using plotters backend
Benchmarking split/position/211: Collecting 100 samples in estimated 5.0054 s (2.5M iter
split/position/211      time:   [1.9883 us 1.9974 us 2.0144 us]
                        change: [-3.5973% -2.1547% -0.8611%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 15 outliers among 100 measurements (15.00%)
  3 (3.00%) low mild
  2 (2.00%) high mild
  10 (10.00%) high severe
Benchmarking split/position/384: Collecting 100 samples in estimated 5.0007 s (2.3M iter
split/position/384      time:   [2.1072 us 2.1113 us 2.1163 us]
                        change: [-0.9981% +0.0554% +1.4612%] (p = 0.92 > 0.05)
                        No change in performance detected.
Found 15 outliers among 100 measurements (15.00%)
  6 (6.00%) high mild
  9 (9.00%) high severe
Benchmarking split/position/557: Collecting 100 samples in estimated 5.0043 s (2.3M iter
split/position/557      time:   [2.1354 us 2.1379 us 2.1409 us]
                        change: [-1.0412% -0.0570% +0.8205%] (p = 0.91 > 0.05)
                        No change in performance detected.
Found 12 outliers among 100 measurements (12.00%)
  12 (12.00%) high severe
Benchmarking split/position/730: Collecting 100 samples in estimated 5.0058 s (2.5M iter
split/position/730      time:   [2.0094 us 2.0162 us 2.0225 us]
                        change: [-2.1488% -1.2633% -0.4645%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe
Benchmarking split/position/903: Collecting 100 samples in estimated 5.0002 s (2.7M iter
split/position/903      time:   [1.8198 us 1.8256 us 1.8320 us]
                        change: [-2.6589% -1.6401% -0.5561%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

Benchmarking split_left/position/211: Collecting 100 samples in estimated 5.0044 s (4.8M
split_left/position/211 time:   [1.0339 us 1.0366 us 1.0395 us]
                        change: [-1.5523% -0.8465% +0.0024%] (p = 0.03 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) high mild
  5 (5.00%) high severe
Benchmarking split_left/position/384: Collecting 100 samples in estimated 5.0019 s (5.5M
split_left/position/384 time:   [903.02 ns 905.73 ns 908.45 ns]
                        change: [-2.7927% -2.1694% -1.4930%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) low mild
  3 (3.00%) high mild
  2 (2.00%) high severe
Benchmarking split_left/position/557: Collecting 100 samples in estimated 5.0006 s (5.2M
split_left/position/557 time:   [956.12 ns 962.81 ns 968.98 ns]
                        change: [-2.2111% -1.4205% -0.6506%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high severe
Benchmarking split_left/position/730: Collecting 100 samples in estimated 5.0026 s (5.7M
split_left/position/730 time:   [872.18 ns 876.23 ns 880.10 ns]
                        change: [-5.5371% -3.8150% -2.1652%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) high mild
  5 (5.00%) high severe
Benchmarking split_left/position/903: Collecting 100 samples in estimated 5.0015 s (6.5M
split_left/position/903 time:   [766.09 ns 790.19 ns 828.75 ns]
                        change: [+0.1781% +1.6384% +3.5133%] (p = 0.05 < 0.05)
                        Change within noise threshold.
Found 10 outliers among 100 measurements (10.00%)
  2 (2.00%) high mild
  8 (8.00%) high severe

Benchmarking split_right/position/211: Collecting 100 samples in estimated 5.0029 s (5.3
split_right/position/211
                        time:   [915.47 ns 921.05 ns 926.75 ns]
                        change: [-4.6356% -3.7931% -3.0282%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
Benchmarking split_right/position/384: Collecting 100 samples in estimated 5.0036 s (4.4
split_right/position/384
                        time:   [1.1490 us 1.1518 us 1.1545 us]
                        change: [-10.391% -7.7376% -5.2257%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  6 (6.00%) high mild
  3 (3.00%) high severe
Benchmarking split_right/position/557: Collecting 100 samples in estimated 5.0025 s (4.3
split_right/position/557
                        time:   [1.1577 us 1.1590 us 1.1605 us]
                        change: [-3.4434% -1.4709% +0.2135%] (p = 0.13 > 0.05)
                        No change in performance detected.
Found 14 outliers among 100 measurements (14.00%)
  4 (4.00%) high mild
  10 (10.00%) high severe
Benchmarking split_right/position/730: Collecting 100 samples in estimated 5.0019 s (4.2
split_right/position/730
                        time:   [1.1777 us 1.1800 us 1.1825 us]
                        change: [-3.1413% -2.2761% -1.5495%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 12 outliers among 100 measurements (12.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
  6 (6.00%) high severe
Benchmarking split_right/position/903: Collecting 100 samples in estimated 5.0010 s (4.7
split_right/position/903
                        time:   [1.0563 us 1.0575 us 1.0590 us]
                        change: [-3.9729% -3.0880% -2.2479%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 16 outliers among 100 measurements (16.00%)
  5 (5.00%) high mild
  11 (11.00%) high severe
```

aslpavel commented:
LGTM, Thank you! I will merge it
